### PR TITLE
enable dns server on gateways

### PR DESCRIPTION
**Pull request by ramaraochavali**
**Please provide a description of this PR:**

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Hmm.. do we want an opt in? I suppose we have `a.cfg.DNSCapture` but 99% of users set this at the mesh config level so applies to everything. I suppose as-is this PR has an opt-out which is maybe enough; the DNS server probably doesn't hurt much beyond a small increase in XDS load?

**Comment by ramaraochavali** [comment]
> the DNS server probably doesn't hurt much beyond a small increase in XDS load?

Yes. And it has an opt-out feature already.

**Comment by howardjohn** [comment]
The test failure is interesting, this change basically means you cannot run a sidecar/ztunnel alongside the gateway

File: pkg/istio-agent/agent.go [hunk 1]
```diff
@@ -536,8 +536,7 @@ func (a *Agent) startFileWatcher(ctx context.Context, filePath string, handler f
 }
 
 func (a *Agent) initLocalDNSServer() (err error) {
-	// we don't need dns server on gateways
-	if a.cfg.DNSCapture && a.cfg.ProxyType == model.SidecarProxy {
+	if a.cfg.DNSCapture {
```
Comment (hzxuzhonghu) [inline pkg/istio-agent/agent.go:539]: we need a separate flag to explicitly allow it on gateway

Comment (hzxuzhonghu) [reply pkg/istio-agent/agent.go:539]: BY this most users make use of dns proxy may run dns proxy in gateway unexpectedly

**Comment by ramaraochavali** [comment]
> The test failure is interesting, this change basically means you cannot run a sidecar/ztunnel alongside the gateway

It is interesting. Now I am convinced that we need a separate opt-in flag for this

File: pilot/cmd/pilot-agent/options/options.go [hunk 2]
```diff
@@ -102,6 +102,9 @@ var (
 	// DNSCaptureByAgent is a copy of the env var in the init code.
 	DNSCaptureByAgent = env.Register("ISTIO_META_DNS_CAPTURE", false,
 		"If set to true, enable the capture of outgoing DNS packets on port 53, redirecting to istio-agent on :15053")
+	// EnableDNSAtGateway enables DNS server at Gateways.
+	EnableDNSAtGateway = env.Register("ISTIO_META_ENABLE_DNS_SERVER", false,
+		"If set to true, starts the DNS server on :15053. This can be used when we want Gateways to use DNS proxy")
```
Comment (hzxuzhonghu) [inline pilot/cmd/pilot-agent/options/options.go:107]: This maybe confusing, by default gateway cannot use it. We have no iptables set for it, right

Comment (ramaraochavali) [reply pilot/cmd/pilot-agent/options/options.go:107]: That is why I called this `ISTIO_META_ENABLE_DNS_SERVER` - so it just starts the DNS server - it does not capture the DNS traffic whoever wants (Dynamic Forward proxy in our example) can point to this resolver

Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/options/options.go:107]: I mean you can mention in the desription

Comment (ramaraochavali) [reply pilot/cmd/pilot-agent/options/options.go]: ok. .updated. PTAL

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**